### ISSUE TITLE
Some Codable extensions generate a compiler crash when building DocC documentation.

### ISSUE THREAD

**Issue by fibrechannelscsi** [apple/swift#62936]
**Description**
When a Codable extension is added to a struct, the option to build DocC documentation in Xcode via `Product -> Build Documentation` no longer works. Instead, a compiler crash is generated:

`Assertion failed: (!empty()), function back, file SmallVector.h, line 291.`

**Steps to reproduce**
In Xcode, paste the following code into a new project:
```
import Foundation
public struct A{}
extension A: Codable {
    public enum CodingKeys: CodingKey {}
    public func encode(to encoder: Encoder) throws {}
    public init(from decoder: Decoder) throws {self.init()}
}
```
Then select `Product -> Build Documentation`.

**Expected behavior**
The documentation should build successfully. A new window displaying the built documentation should appear once the process is complete.

**Environment**
The compiler crash occurs with toolchains 2022-09-27a or later. However, this builds successfully under 2022-08-15a or earlier. (I haven't narrowed down the date range any further than this at this point.)
Xcode 13.3, Build version 13E113
Deployment target: M1 (Fails on x86 as well, same version of Xcode.)

Stack trace:
```
1.	Apple Swift version 5.8-dev (LLVM ed4d811c2897d77, Swift 8491f52d478aca6)
2.	Compiling with the current language version
3.	While evaluating request ExecuteSILPipelineRequest(Run pipelines { Non-Diagnostic Mandatory Optimizations, Serialization, Rest of Onone } on SIL for someProject)
4.	While running pass #56 SILModuleTransform "SerializeSILPass".
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x0000000109b4ec3c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  swift-frontend           0x0000000109b4dec0 llvm::sys::RunSignalHandlers() + 128
2  swift-frontend           0x0000000109b4f27c SignalHandler(int) + 304
3  libsystem_platform.dylib 0x000000018d1314c4 _sigtramp + 56
4  libsystem_pthread.dylib  0x000000018d119ee0 pthread_kill + 288
5  libsystem_c.dylib        0x000000018d054340 abort + 168
6  libsystem_c.dylib        0x000000018d053754 err + 0
7  swift-frontend           0x0000000109b59c64 swift::symbolgraphgen::SymbolGraphASTWalker::walkToDeclPre(swift::Decl*, swift::CharSourceRange)::$_1::operator()(swift::Type) const (.cold.5) + 0
8  swift-frontend           0x00000001052ff404 swift::symbolgraphgen::SymbolGraphASTWalker::walkToDeclPre(swift::Decl*, swift::CharSourceRange)::$_1::operator()(swift::Type) const + 852
9  swift-frontend           0x00000001052feea4 swift::symbolgraphgen::SymbolGraphASTWalker::walkToDeclPre(swift::Decl*, swift::CharSourceRange) + 1020
10 swift-frontend           0x000000010550c128 (anonymous namespace)::SemaAnnotator::walkToDeclPre(swift::Decl*) + 1760
11 swift-frontend           0x00000001063b9928 (anonymous namespace)::Traversal::doIt(swift::Decl*) + 200
12 swift-frontend           0x00000001063b9854 swift::Decl::walk(swift::ASTWalker&) + 32
13 swift-frontend           0x000000010550a774 swift::SourceEntityWalker::walk(swift::Decl*) + 192
14 swift-frontend           0x00000001052fdb70 swift::symbolgraphgen::emitSymbolGraphForModule(swift::ModuleDecl*, swift::symbolgraphgen::SymbolGraphOptions const&) + 424
15 swift-frontend           0x00000001057b7288 swift::serialize(llvm::PointerUnion<swift::ModuleDecl*, swift::SourceFile*>, swift::SerializationOptions const&, swift::symbolgraphgen::SymbolGraphOptions const&, swift::SILModule const*, swift::fine_grained_dependencies::SourceFileDepGraph const*) + 340
16 swift-frontend           0x00000001052cf134 bool llvm::function_ref<bool (swift::fine_grained_dependencies::SourceFileDepGraph&&)>::callback_fn<performCompileStepsPostSILGen(swift::CompilerInstance&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::PointerUnion<swift::ModuleDecl*, swift::SourceFile*>, swift::PrimarySpecificPaths const&, int&, swift::FrontendObserver*)::$_26::operator()() const::'lambda'(swift::fine_grained_dependencies::SourceFileDepGraph&&)>(long, swift::fine_grained_dependencies::SourceFileDepGraph&&) + 40
17 swift-frontend           0x000000010652a5e4 swift::fine_grained_dependencies::withReferenceDependencies(llvm::PointerUnion<swift::ModuleDecl const*, swift::SourceFile const*>, swift::DependencyTracker const&, llvm::StringRef, bool, llvm::function_ref<bool (swift::fine_grained_dependencies::SourceFileDepGraph&&)>) + 152
18 swift-frontend           0x00000001052cf0a8 std::__1::__function::__func<performCompileStepsPostSILGen(swift::CompilerInstance&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::PointerUnion<swift::ModuleDecl*, swift::SourceFile*>, swift::PrimarySpecificPaths const&, int&, swift::FrontendObserver*)::$_26, std::__1::allocator<performCompileStepsPostSILGen(swift::CompilerInstance&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::PointerUnion<swift::ModuleDecl*, swift::SourceFile*>, swift::PrimarySpecificPaths const&, int&, swift::FrontendObserver*)::$_26>, void ()>::operator()() + 284
19 swift-frontend           0x00000001053d3240 swift::SILModule::serialize() + 44
20 swift-frontend           0x0000000105c8687c SerializeSILPass::run() + 156
21 swift-frontend           0x0000000105b45060 swift::SILPassManager::runModulePass(unsigned int) + 984
22 swift-frontend           0x0000000105b4ac20 swift::SILPassManager::execute() + 624
23 swift-frontend           0x0000000105b42014 swift::SILPassManager::executePassPipelinePlan(swift::SILPassPipelinePlan const&) + 68
24 swift-frontend           0x0000000105b41f9c swift::ExecuteSILPipelineRequest::evaluate(swift::Evaluator&, swift::SILPipelineExecutionDescriptor) const + 68
25 swift-frontend           0x0000000105b65780 swift::SimpleRequest<swift::ExecuteSILPipelineRequest, std::__1::tuple<> (swift::SILPipelineExecutionDescriptor), (swift::RequestFlags)1>::evaluateRequest(swift::ExecuteSILPipelineRequest const&, swift::Evaluator&) + 28
26 swift-frontend           0x0000000105b4df38 llvm::Expected<swift::ExecuteSILPipelineRequest::OutputType> swift::Evaluator::getResultUncached<swift::ExecuteSILPipelineRequest>(swift::ExecuteSILPipelineRequest const&) + 252
27 swift-frontend           0x0000000105b42210 swift::executePassPipelinePlan(swift::SILModule*, swift::SILPassPipelinePlan const&, bool, swift::irgen::IRGenModule*) + 84
28 swift-frontend           0x0000000105b529b0 swift::runSILPassesForOnone(swift::SILModule&) + 72
29 swift-frontend           0x00000001053299ec swift::CompilerInstance::performSILProcessing(swift::SILModule*) + 240
30 swift-frontend           0x00000001052cb520 performCompileStepsPostSILGen(swift::CompilerInstance&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::PointerUnion<swift::ModuleDecl*, swift::SourceFile*>, swift::PrimarySpecificPaths const&, int&, swift::FrontendObserver*) + 716
31 swift-frontend           0x00000001052caebc swift::performCompileStepsPostSema(swift::CompilerInstance&, int&, swift::FrontendObserver*) + 1100
32 swift-frontend           0x00000001052d9a5c withSemanticAnalysis(swift::CompilerInstance&, swift::FrontendObserver*, llvm::function_ref<bool (swift::CompilerInstance&)>, bool) + 160
33 swift-frontend           0x00000001052cc9d4 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3256
34 swift-frontend           0x0000000105135828 swift::mainEntry(int, char const**) + 3136
35 dyld                     0x0000000113161088 start + 516
```

**Additional information**
- Non-empty structs also exhibit this behavior, as well as fully-populated `encode()` and `decode()` functions.
- Moving the CodingKeys, encode and decode functionality out of the extension and into the struct itself allows the compilation to succeed on all toolchains tested.

**Comment by Kyle-Ye** [comment]
I can reproduce the bug on

**Comment by fibrechannelscsi** [comment]
I've narrowed down the toolchains a bit more: it looks like this is broken on 2023-01-02a, but was fixed as of 2023-01-07a. 
Thanks!

**Comment by Kyle-Ye** [comment]
This is fixed by #62825
